### PR TITLE
KAFKA-14647: Moving TopicFilter to server-common/utils

### PR DESCRIPTION
**Pull request by vamossagar12**
Moving TopicFilter to server-common/utils

### CODE REVIEW

**Comment by vamossagar12** [comment]
@fvaleri , plz review this PR.

File: server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                .replaceAll("^[\"']+", "")
+                .replaceAll("[\"']+$", ""); // property files may bring quotes
+        try {
+            Pattern.compile(regex);
+        } catch (PatternSyntaxException e) {
+            throw new RuntimeException(regex + " is an invalid regex.");
+        }
+    }
+
+    public abstract boolean isTopicAllowed(String topic, boolean excludeInternalTopics);
```
Comment (fvaleri) [inline server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java:44]: It's probably useful to add an overloaded method that only takes the topic name and excludes internal topics by default. This would also simplify tests.

File: server-common/src/test/java/org/apache/kafka/server/util/TopicFilterTest.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+        IncludeList topicFilter1 = new IncludeList("yes1,yes2");
+        assertTrue(topicFilter1.isTopicAllowed("yes2", true));
+        assertTrue(topicFilter1.isTopicAllowed("yes2", false));
+        assertFalse(topicFilter1.isTopicAllowed("no1", true));
+        assertFalse(topicFilter1.isTopicAllowed("no1", false));
+
+        IncludeList topicFilter2 = new IncludeList(".+");
+        assertTrue(topicFilter2.isTopicAllowed("alltopics", true));
+        assertFalse(topicFilter2.isTopicAllowed(Topic.GROUP_METADATA_TOPIC_NAME, true));
```
Comment (fvaleri) [inline server-common/src/test/java/org/apache/kafka/server/util/TopicFilterTest.java:37]: What about the other internal topic? You can use the INTERNAL_TOPICS set and test all of them.

File: server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.server.util;
+
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+
+public abstract class TopicFilter {
```
Comment (fvaleri) [inline server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java:26]: We cannot replicate the sealed behavior until Java 17, but we can have a single file containing all implementations like this:

```sh
public interface TopicFilter {
    boolean isTopicAllowed(String topic, boolean excludeInternalTopics);
    
    public static class IncludeList implements TopicFilter {
      // ...
    }
}
```

File: server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    public abstract boolean isTopicAllowed(String topic, boolean excludeInternalTopics);
+
+    @Override
+    public String toString() {
+        return this.regex;
+    }
+
+    public String getRegex() {
```
Comment (fvaleri) [inline server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java]: Why we need this method? I think we can remove it.

File: server-common/src/main/java/org/apache/kafka/server/util/IncludeList.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package org.apache.kafka.server.util;
+
+import org.apache.kafka.common.internals.Topic;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class IncludeList extends TopicFilter {
+
```
Comment (fvaleri) [inline server-common/src/main/java/org/apache/kafka/server/util/IncludeList.java]: Please, remove this extra line.

File: server-common/src/test/java/org/apache/kafka/server/util/TopicFilterTest.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+package org.apache.kafka.server.util;
+
+import org.apache.kafka.common.internals.Topic;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+public class TopicFilterTest {
+
```
Comment (fvaleri) [inline server-common/src/test/java/org/apache/kafka/server/util/TopicFilterTest.java]: Please, remove this extra line and all the others at the start/end of classes/methods.

**Review by fvaleri — CHANGES_REQUESTED** [review]
Hi @vamossagar12, thanks for working on this! 

I left some comments, but there are build errors.

In this PR, it is probably a good idea to also move the TopicPartitionFilter and related code contained in GetOffsetShell. They are very similar and of general use.

File: server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java (hunk 1 above)
Comment (vamossagar12) [reply server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java:44]: I thought about this and I feel retaining the current scheme seems to be better as the only 2 possible states of inclusion/exclusion can be control by the single flag.

File: server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java (hunk 3 above)
Comment (vamossagar12) [reply server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java:26]: I had thought about this at the time of implementation. The reason I didn't do this was that the original implementation used `sealed abstract class` which probably helps in exhaustive pattern matching. Also, I am not sure how would having the classes in the same file help in this case. Maybe I am missing something?

File: server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java (hunk 4 above)
Comment (vamossagar12) [reply server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java]: Oh actually MirrorMaker.scala file needed access to the regex object. I thought exposing it via a getter might be a better idea instead of exposing the field itself. Can remove if you think it isn't needed.

**Comment by vamossagar12** [comment]
> TopicPartitionFilter

@fvaleri , sorry about the build failures. I hadn't run the checkstyle for core on my local before pushing. I will work on the TopicPartitionFilter class, rest of the comments have been addressed/answered. Thanks for reviewing.

File: server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java (hunk 3 above)
Comment (fvaleri) [reply server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java:26]: It helps with reducing the number of source files and clarity as you have all implementations on the same page. This is similar to what the original scala code and TopicPartitionFilter are doing.

File: server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java (hunk 4 above)
Comment (fvaleri) [reply server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java]: Ok, you could simply return a Pattern in this case.

File: server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java (hunk 1 above)
Comment (fvaleri) [reply server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java:44]: Fair enough.

File: server-common/src/test/java/org/apache/kafka/server/util/TopicFilterTest.java (hunk 2 above)
Comment (vamossagar12) [reply server-common/src/test/java/org/apache/kafka/server/util/TopicFilterTest.java:37]: `INTERNAL_TOPICS` variables is private so I didn't use it directly. But I added the test for the other internal topic.

File: server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java (hunk 4 above)
Comment (vamossagar12) [reply server-common/src/main/java/org/apache/kafka/server/util/TopicFilter.java]: That's a good idea. Did that. Would be pushing in some time.

**Comment by vamossagar12** [comment]
Thanks @fvaleri , I made the changes.

File: checkstyle/import-control.xml [hunk 7]
```diff
@@ -351,6 +351,7 @@
 
     <!-- This is required to make AlterConfigPolicyTest work. -->
     <allow pkg="org.apache.kafka.server.policy" />
+    <allow pkg="org.apache.kafka.server.util" />
```
Comment (fvaleri) [inline checkstyle/import-control.xml]: Is this related to the above comment? If not please move it above right after joptsimple.

File: server-common/src/main/java/org/apache/kafka/server/util/TopicPartitionFilter.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.server.util.TopicFilter.IncludeList;
+
+import java.util.List;
+
+public interface TopicPartitionFilter {
+    /**
+     * Used to filter topics based on a certain criteria, for example, a set of topic names or a regular expression.
+     */
+    boolean isTopicAllowed(String topic);
```
Comment (fvaleri) [inline server-common/src/main/java/org/apache/kafka/server/util/TopicPartitionFilter.java:28]: Add an empty line after this.

File: server-common/src/main/java/org/apache/kafka/server/util/TopicPartitionFilter.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        public boolean isTopicAllowed(String topic) {
+            return filters.stream().anyMatch(tp -> tp.isTopicAllowed(topic));
+        }
+
+        @Override
+        public boolean isTopicPartitionAllowed(TopicPartition partition) {
+            return filters.stream().anyMatch(tp -> tp.isTopicPartitionAllowed(partition));
+        }
+    }
+
```
Comment (fvaleri) [inline server-common/src/main/java/org/apache/kafka/server/util/TopicPartitionFilter.java]: Remove empty line.

File: server-common/src/test/java/org/apache/kafka/server/util/TopicFilterTest.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        IncludeList topicFilter3 = new IncludeList("included-topic.+");
+        assertTrue(topicFilter3.isTopicAllowed("included-topic1", true));
+        assertFalse(topicFilter3.isTopicAllowed("no1", true));
+
+        IncludeList topicFilter4 = new IncludeList("test-(?!bad\\b)[\\w]+");
+        assertTrue(topicFilter4.isTopicAllowed("test-good", true));
+        assertFalse(topicFilter4.isTopicAllowed("test-bad", true));
+    }
+
```
Comment (fvaleri) [inline server-common/src/test/java/org/apache/kafka/server/util/TopicFilterTest.java]: Remove empty line.

**Review by fvaleri — COMMENTED** [review]
Hi @vamossagar12, I left few comments, but looks mostly good. 

I guess we also need to add some unit tests for PartitionFilter and TopicPartitionFilter, like we have for TopicFilter. They are in GetOffsetShellParsingTest, which should rather test the tool interface not filters.

**Comment by vamossagar12** [comment]
Thanks @fvaleri , I addressed the review comments. Regarding unit tests for PartitionFilter and TopicPartitionFilter, I checked the tests in `GetOffsetShellParsingTest` and they are pretty comprehensive. I also think the interface per se is being tested in `GetOffsetShellTest` so I think we can let the tests remain in `GetOffsetShellParsingTest`. LMKWYT.

**Comment by fvaleri** [comment]
Hi @vamossagar12 this looks good, but I still think we should move `GetOffsetShellParsingTest` to `TopicPartitionFilterTest` removing any reference to `GetOffsetShell` (there is already a test for that) and create a new `PartitionFilterTest`, which should be similar to `TopicFilterTest`.

**Comment by vamossagar12** [comment]
> Hi @vamossagar12 this looks good, but I still think we should move `GetOffsetShellParsingTest` to `TopicPartitionFilterTest` removing any reference to `GetOffsetShell` (there is already a test for that) and create a new `PartitionFilterTest`, which should be similar to `TopicFilterTest`.

Thanks @fvaleri . I tried doing this today. A problem that I see is that `GetOffsetShellParsingTest` makes use of 2 methods from `GetOffsetShell` namely `createTopicPartitionFilterWithTopicAndPartitionPattern` and `createTopicPartitionFilterWithPatternList`. Those 2 methods are also eventually used in `GetOffsetShell` when creating `topicPartitionFilter`. Ideally these 2 methods can reside in `ToolsUtils` and have both tests and `GetOffsetShell` reference it. 
I see there's also [Move GetOffsetShell to tools](https://issues.apache.org/jira/browse/KAFKA-14581). Can we do the move as part of that effort or here? WDYS?

**Comment by ijuma** [comment]
What is the justification for this? For something to be in server-common, it needs to be used by multiple modules.

**Comment by fvaleri** [comment]
@ijuma these classes are currently used in both tools and core modules. Not all tools will be migrated in the same release and we also have the deprecated MirrorMaker 1, which we do not plan to move.

**Comment by vamossagar12** [comment]
Sorry I completely missed this. Thanks @fvaleri . @ijuma do you agree with the reasoning provided by Federico?

**Comment by dengziming** [comment]
TopicFilter and TopicPartitionFilter are very customized and simple interfaces, it's unnecessary to rewrite them currently IMO, at least currently, they are only used in a few tools, especially `TopicPartitionFilter` which are only used in GetOffsetShell.

**Comment by vamossagar12** [comment]
In that case, would it make sense to close this PR @fvaleri ?

**Comment by fvaleri** [comment]
> In that case, would it make sense to close this PR @fvaleri ?

Why? We can discuss where to place these interfaces, but I would keep the work that has been done. It's just another step towards KAFKA-14524.

**Comment by vamossagar12** [comment]
I see. TBH I haven't followed that migration effort off-late. I think there are build failures as well which I haven't addressed yet.

**Comment by ruslankrivoshein** [comment]
> I would keep the work that has been done

I agree. This work is already in use in another PR.

**Comment by vamossagar12** [comment]
> > I would keep the work that has been done
> 
> I agree. This work is already in use in another PR.

I see. In that case, I need to fix the build issue and ask folks for reviews then. Will try to do so as soon as I can. Thanks for notifying.

**Comment by vamossagar12** [comment]
@ruslankrivoshein , I have fixed the checkstyle issues. Also, I believe that the other comment [here](https://github.com/apache/kafka/pull/13158#issuecomment-1422555044) has been addressed by you in the PR https://github.com/apache/kafka/pull/13562. 
Sorry for the long wait here- it just went off my radar. @fvaleri , do you think this is looking fine now?

**Comment by fvaleri** [comment]
@vamossagar12 LGTM. Thanks.

**Comment by vamossagar12** [comment]
Thanks @fvaleri . Hmm I see 101 test failures. 92 existing and 9 new. Atleast the new ones look unrelated..

**Comment by mimaison** [comment]
I took a look at the changes and they look fine. If I understand correctly we're moving these classes to server-common while we have to keep MirrorMaker. Then in Kafka 4.0 (and assuming [KAFKA-14525](https://issues.apache.org/jira/browse/KAFKA-14525) is complete) we will be able to move all these classes to tools. Is that right?

**Comment by fvaleri** [comment]
Yes, once the MirrorMaker1 dependency will be gone in 4.0.0, we can move them to tools. I added a note in [KAFKA-14705](https://issues.apache.org/jira/browse/KAFKA-14705).

**Comment by vamossagar12** [comment]
> I took a look at the changes and they look fine. If I understand correctly we're moving these classes to server-common while we have to keep MirrorMaker. Then in Kafka 4.0 (and assuming [KAFKA-14525](https://issues.apache.org/jira/browse/KAFKA-14525) is complete) we will be able to move all these classes to tools. Is that right?

Thanks @mimaison , yes that's the idea.

**Comment by vamossagar12** [comment]
@mimaison can this be merged if all changes look fine?

**Review by mimaison — APPROVED** [review]
Thanks!
